File: src/cli/commands/diff.ts
import { resolve } from 'node:path';
import { existsSync } from 'node:fs';
import chalk from 'chalk';
import { GitBridge } from '../../git/bridge.js';
import type { DiffScope } from '../../git/types.js';
import { ParserRegistry } from '../../parser/registry.js';
import { computeSemanticDiff } from '../../parser/differ.js';
import { SemDatabase } from '../../storage/database.js';
import { formatTerminal } from '../formatters/terminal.js';
import { formatJson } from '../formatters/json.js';
import { createDefaultRegistry } from '../../parser/plugins/index.js';
import { loadConfig, validateChanges, formatValidationResults } from './validate.js';

export interface DiffOptions {
  cwd?: string;
  format?: 'terminal' | 'json';
  staged?: boolean;
  commit?: string;
  from?: string;
  to?: string;
  store?: boolean;
}

// Singleton registry — no need to recreate on every call
let _registry: ParserRegistry | undefined;
function getRegistry(): ParserRegistry {
  return (_registry ??= createDefaultRegistry());
}

export async function diffCommand(opts: DiffOptions = {}): Promise<void> {
  const cwd = opts.cwd ?? process.cwd();
  const git = new GitBridge(cwd);

  let scope: DiffScope;
  let fileChanges;

  // Fast path: auto-detect uses combined detectAndGetFiles (1 round of git calls)
  // Explicit scope falls back to separate calls
  if (opts.commit) {
    if (!(await git.isRepo())) { console.error(chalk.red('Error: Not inside a Git repository.')); process.exit(1); }
    scope = { type: 'commit', sha: opts.commit };
    fileChanges = await git.getChangedFiles(scope);
  } else if (opts.from && opts.to) {
    if (!(await git.isRepo())) { console.error(chalk.red('Error: Not inside a Git repository.')); process.exit(1); }
    scope = { type: 'range', from: opts.from, to: opts.to };
    fileChanges = await git.getChangedFiles(scope);
  } else if (opts.staged) {
    if (!(await git.isRepo())) { console.error(chalk.red('Error: Not inside a Git repository.')); process.exit(1); }
    scope = { type: 'staged' };
    fileChanges = await git.getChangedFiles(scope);
  } else {
    // Combined: isRepo + detectScope + getChangedFiles in one batch
    try {
      const result = await git.detectAndGetFiles();
      scope = result.scope;
      fileChanges = result.files;
    } catch {
      console.error(chalk.red('Error: Not inside a Git repository.'));
      process.exit(1);
    }
  }

  if (fileChanges.length === 0) {
    console.log(chalk.dim('No changes detected.'));
    return;
  }

  // Compute semantic diff
  const registry = getRegistry();
  const commitSha = scope.type === 'commit' ? scope.sha : undefined;
  const result = computeSemanticDiff(fileChanges, registry, commitSha);

  // Get repoRoot once for both store + validation
  const repoRoot = await git.getRepoRoot();

  // Optionally store changes
  if (opts.store) {
    const dbPath = resolve(repoRoot, '.sem', 'sem.db');
    if (existsSync(dbPath)) {
      const db = new SemDatabase(dbPath);
      db.insertChanges(result.changes);
      db.close();
    }
  }

  // Output
  const format = opts.format ?? 'terminal';
  if (format === 'json') {
    console.log(formatJson(result));
  } else {
    console.log(formatTerminal(result));
  }

  // Run validation rules if .semrc exists
  try {
    const config = await loadConfig(repoRoot);
    if (config.rules && config.rules.length > 0) {
      const violations = validateChanges(result, config);
      if (violations.length > 0) {
        console.log('');
        console.log(formatValidationResults(violations));
      }
    }
  } catch {
    // No config or invalid config — skip validation
  }
}

File: crates/sem-cli/src/commands/diff.rs
use std::io::Read;
use std::path::Path;
use std::process;
use std::time::Instant;

use sem_core::git::bridge::GitBridge;
use sem_core::git::types::{DiffScope, FileChange};
use sem_core::parser::differ::compute_semantic_diff;
use sem_core::parser::plugins::create_default_registry;

use crate::formatters::{json::format_json, terminal::format_terminal};

pub struct DiffOptions {
    pub cwd: String,
    pub format: OutputFormat,
    pub staged: bool,
    pub commit: Option<String>,
    pub from: Option<String>,
    pub to: Option<String>,
    pub stdin: bool,
    pub profile: bool,
    pub file_exts: Vec<String>,
    pub files: Vec<String>,
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum OutputFormat {
    Terminal,
    Json,
}

pub fn diff_command(opts: DiffOptions) {
    let total_start = Instant::now();

    let t0 = Instant::now();
    let (file_changes, from_stdin) = if opts.files.len() == 2 {
        // Compare two arbitrary files: sem diff file1.ts file2.ts
        let path_a = Path::new(&opts.files[0]);
        let path_b = Path::new(&opts.files[1]);

        let content_a = std::fs::read_to_string(path_a).unwrap_or_else(|e| {
            eprintln!("\x1b[31mError reading {}: {e}\x1b[0m", path_a.display());
            process::exit(1);
        });
        let content_b = std::fs::read_to_string(path_b).unwrap_or_else(|e| {
            eprintln!("\x1b[31mError reading {}: {e}\x1b[0m", path_b.display());
            process::exit(1);
        });

        let change = FileChange {
            file_path: opts.files[1].clone(),
            old_file_path: None,
            status: sem_core::git::types::FileStatus::Modified,
            before_content: Some(content_a),
            after_content: Some(content_b),
        };
        (vec![change], false)
    } else if opts.files.len() == 1 {
        eprintln!("\x1b[31mError: provide two files to compare, or none for git diff.\x1b[0m");
        process::exit(1);
    } else if opts.stdin {
        // Read FileChange[] from stdin — no git repo needed
        let mut input = String::new();
        std::io::stdin().read_to_string(&mut input).unwrap_or_else(|e| {
            eprintln!("\x1b[31mError reading stdin: {e}\x1b[0m");
            process::exit(1);
        });
        let changes: Vec<FileChange> = serde_json::from_str(&input).unwrap_or_else(|e| {
            eprintln!("\x1b[31mError parsing stdin JSON: {e}\x1b[0m");
            process::exit(1);
        });
        (changes, true)
    } else {
        let git = match GitBridge::open(Path::new(&opts.cwd)) {
            Ok(g) => g,
            Err(_) => {
                eprintln!("\x1b[31mError: Not inside a Git repository.\x1b[0m");
                process::exit(1);
            }
        };

        let (_scope, file_changes) = if let Some(ref sha) = opts.commit {
            let scope = DiffScope::Commit { sha: sha.clone() };
            match git.get_changed_files(&scope) {
                Ok(files) => (scope, files),
                Err(e) => {
                    eprintln!("\x1b[31mError: {e}\x1b[0m");
                    process::exit(1);
                }
            }
        } else if let (Some(ref from), Some(ref to)) = (&opts.from, &opts.to) {
            let scope = DiffScope::Range {
                from: from.clone(),
                to: to.clone(),
            };
            match git.get_changed_files(&scope) {
                Ok(files) => (scope, files),
                Err(e) => {
                    eprintln!("\x1b[31mError: {e}\x1b[0m");
                    process::exit(1);
                }
            }
        } else if opts.staged {
            let scope = DiffScope::Staged;
            match git.get_changed_files(&scope) {
                Ok(files) => (scope, files),
                Err(e) => {
                    eprintln!("\x1b[31mError: {e}\x1b[0m");
                    process::exit(1);
                }
            }
        } else {
            match git.detect_and_get_files() {
                Ok((scope, files)) => (scope, files),
                Err(_) => {
                    eprintln!("\x1b[31mError: Not inside a Git repository.\x1b[0m");
                    process::exit(1);
                }
            }
        };
        (file_changes, false)
    };
    let git_diff_ms = t0.elapsed().as_secs_f64() * 1000.0;

    // Filter by file extensions if specified
    let file_changes = if opts.file_exts.is_empty() {
        file_changes
    } else {
        let exts: Vec<String> = opts.file_exts.iter().map(|e| {
            if e.starts_with('.') { e.clone() } else { format!(".{}", e) }
        }).collect();
        file_changes.into_iter().filter(|fc| {
            exts.iter().any(|ext| fc.file_path.ends_with(ext.as_str()))
        }).collect()
    };

    if file_changes.is_empty() {
        println!("\x1b[2mNo changes detected.\x1b[0m");
        return;
    }

    let t2 = Instant::now();
    let registry = create_default_registry();
    let registry_ms = t2.elapsed().as_secs_f64() * 1000.0;

    let t3 = Instant::now();
    let result = compute_semantic_diff(&file_changes, &registry, None, None);
    let parse_diff_ms = t3.elapsed().as_secs_f64() * 1000.0;

    let t4 = Instant::now();
    let output = match opts.format {
        OutputFormat::Json => format_json(&result),
        OutputFormat::Terminal => format_terminal(&result),
    };
    let format_ms = t4.elapsed().as_secs_f64() * 1000.0;

    println!("{output}");

    if opts.profile {
        let total_ms = total_start.elapsed().as_secs_f64() * 1000.0;
        eprintln!();
        eprintln!("\x1b[2m── Profile ──────────────────────────────────\x1b[0m");
        eprintln!("\x1b[2m  input ({})  {git_diff_ms:>8.2}ms\x1b[0m",
            if from_stdin { "stdin" } else { "git" });
        eprintln!("\x1b[2m  registry init        {registry_ms:>8.2}ms\x1b[0m");
        eprintln!("\x1b[2m  parse + match        {parse_diff_ms:>8.2}ms\x1b[0m");
        eprintln!("\x1b[2m  format output        {format_ms:>8.2}ms\x1b[0m");
        eprintln!("\x1b[2m  ─────────────────────────────────\x1b[0m");
        eprintln!("\x1b[2m  total                {total_ms:>8.2}ms\x1b[0m");
        eprintln!("\x1b[2m  files: {}  entities: {}  changes: {}\x1b[0m",
            file_changes.len(), result.changes.len(),
            result.added_count + result.modified_count + result.deleted_count + result.moved_count + result.renamed_count);
        eprintln!("\x1b[2m─────────────────────────────────────────────\x1b[0m");
    }
}
